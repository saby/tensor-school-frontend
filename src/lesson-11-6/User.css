.User {
    display: flex;
    flex-direction: column;
    position: relative;
}
.User__fio {
    font-weight: bold;
    font-size: 2rem;
    flex-shrink: 0;
}
.User__additionalInfo {
    text-align: left;
}
.User__tabs {
    display: flex;
    width: 100%;
    box-shadow: inset 1px -1px 1px lightgrey;
    margin-top: 16px;
}
.User__tab {
    box-shadow: inset 0 -2px 2px lightgray;
    padding: 4px;
}
.User__tab:not(.User__tab-selected) {
    cursor: pointer;
}
.User__tab:not(.User__tab-selected):hover {
    background-color: #05FFFF21;
}
.User__tab-selected {
    box-shadow: inset 0 0 2px lightgray;
    border-bottom: 3px solid #05ffff;
}
.User__info {
    display: flex;
    flex-direction: column;
}
.User__row {
    margin-top: 4px;
    display: flex;
    justify-content: space-between;
}
.User::before {
    content: '';
    position: absolute;
    width: 233px;
    height: 302px;
    background-image: url("./zam.jpg");
    background-size: cover;
    left: -248px;
    top: 0;
    z-index: -1;
}
