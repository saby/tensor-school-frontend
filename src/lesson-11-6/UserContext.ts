import {createContext} from "react";

interface ISchool {
    name: string;
    address: string;
    startDate: Date;
    finishDate: Date;
}

interface IUniversity extends ISchool {
    specialization: string;
}

export interface IUserData {
    firstName: string;
    lastName: string;
    patronym: string;
    gender: 'мужской' | 'женский';
    birthdate: Date;

    phone: string;
    email: string;
    location: string;

    work: string;
    hobby: string;
    music: string;
    films: string;
    books: string;
    games: string;
    about: string;

    school: Partial<ISchool>;
    university: Partial<IUniversity>;
}

interface IUserUpdate {
    updateUser: (data: Partial<IUserData>) => void;
}

export interface IUserContext extends Partial<IUserData>, IUserUpdate {}

export const UserContext = createContext<IUserContext>({updateUser() {}});
