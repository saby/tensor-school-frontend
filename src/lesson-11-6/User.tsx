import {IUserContext, UserContext} from "./UserContext";
import {MouseEventHandler, ReactElement, useCallback, useContext, useState} from "react";
import Contacts from "./Contacts";
import Interests from "./Interests";
import Education from "./Education";
import './User.css'

type Tabs = 'contacts' | 'interests' | 'education';

function Content(tab: Tabs): ReactElement {
    const contents: Record<Tabs, ReactElement> = {
        contacts: <Contacts/>,
        interests: <Interests/>,
        education: <Education/>
    }
    return contents[tab];
}

function Tab({selectedTab, children, id}: { selectedTab: string, children: string, id: string }): ReactElement {
    return (<div data-id={id} className={'User__tab ' + (selectedTab === id ? 'User__tab-selected' : '')}>{children}</div>)
}

function User() {
    const {firstName, lastName, patronym, gender, birthdate} = useContext<IUserContext>(UserContext);
    const [selectedTab, setSelected] = useState<Tabs>('contacts');
    const tabClick = useCallback<MouseEventHandler<HTMLElement>>(({ target }) => {
        const HTMLTarget = (target as HTMLElement);
        if (HTMLTarget.classList.contains('User__tab')) {
            const tab = HTMLTarget.getAttribute('data-id') as Tabs;
            if (tab !== selectedTab) {
                setSelected(tab);
            }
        }
    }, [selectedTab, setSelected])
    return (
        <div className="User">
            <div className="User__fio">{`${firstName} ${lastName} ${patronym}`}</div>
            <div className="User__additionalInfo">{gender} {birthdate?.toLocaleDateString('ru')}</div>
            <div className="User__tabs" onClick={tabClick}>
                <Tab id="contacts" selectedTab={selectedTab}>Контакты</Tab>
                <Tab id="interests" selectedTab={selectedTab}>Интересы</Tab>
                <Tab id="education" selectedTab={selectedTab}>Образование</Tab>
            </div>
            <div className="User__info">
                {Content(selectedTab)}
            </div>
        </div>
    )
}

export default User;
