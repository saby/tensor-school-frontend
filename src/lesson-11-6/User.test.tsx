import {fireEvent, render} from "@testing-library/react";
import {IUserContext, IUserData, UserContext} from "./UserContext";
import User from "./User";

function renderTestContext(customContext?: Partial<IUserContext>) {
    let context: IUserContext = {
        about: "человек",
        birthdate: new Date(1997, 5, 13),
        books: "дюна",
        email: "myemail@mail.com",
        films: "дюна",
        firstName: "Иванов",
        games: "Ведьмак",
        gender: 'мужской',
        hobby: "реверс-инжиниринг",
        lastName: "Иван",
        location: "г. Ярославль",
        music: "mood music",
        patronym: "Иванович",
        phone: "+701234567890",
        school: {
            name: '322',
            address: 'Ярославль, ул. Пушкина д. Колотушкина',
            startDate: new Date(2004, 8),
            finishDate: new Date(2015, 4)
        },
        university: {
            name: 'ЯрГУ им.П.Г.Демидова',
            address: 'Ярославль, ул. Союзная 144',
            startDate: new Date(2015, 8),
            finishDate: new Date(2019, 7, 31),
            specialization: 'Компьютерные науки'
        },
        work: "Тензор",
        updateUser(updateUserData: Partial<IUserData>): void {
            context = { ...context, ...updateUserData };
            rerender(<UserContext.Provider value={context}>
                <User/>
            </UserContext.Provider>);
        },
        ...customContext
    }
    const result = render(
        <UserContext.Provider value={context}>
            <User/>
        </UserContext.Provider>
    )
    const {rerender} = result;

    return result
}

describe('Context', () => {
    it('Вкладка по умолчанию - Контакты', () => {
        const {container} = renderTestContext();
        expect(container).toMatchSnapshot();
    })

    it('Вкладка Интересы', () => {
        const {container, getByText} = renderTestContext();
        fireEvent.click(getByText('Интересы'));
        expect(container).toMatchSnapshot();
    })

    it('Вкладка Образование', () => {
        const {container, getByText} = renderTestContext();
        fireEvent.click(getByText('Образование'));
        expect(container).toMatchSnapshot();
    })

    it('Меняем контакты', () => {
        const {container, getByText} = renderTestContext();

        const matchInput = (id: string, action?: (input: Element) => void) => {
            const input = container.querySelector(`[data-id="${id}"]`) as Element;
            action && action(input);
            expect(input).toMatchSnapshot();
        }

        matchInput('phone', (input) =>
            fireEvent.change(input, { target: { value: '+79876543210' }}));

        matchInput('email', (input) =>
            fireEvent.change(input, { target: { value: 'not_my@email.com' }}));

        matchInput('location', (input) =>
            fireEvent.change(input, { target: { value: 'где-то' }}));

        fireEvent.click(getByText('Образование'));
        fireEvent.click(getByText('Контакты'));

        ['phone', 'email', 'location'].forEach((id) => matchInput(id));
    })

    it('Меняем интересы', () => {
        const {container, getByText} = renderTestContext();
        fireEvent.click(getByText('Интересы'));

        const matchInput = (id: string, action?: (input: Element) => void) => {
            const input = container.querySelector(`[data-id="${id}"]`) as Element;
            action && action(input);
            expect(input).toMatchSnapshot();
        }

        ['work', 'hobby', 'music', 'films', 'books', 'games', 'about'].forEach((id) =>
            matchInput(id, (input) =>
                fireEvent.change(input, { target: { value: '' }})
            )
        );

        fireEvent.click(getByText('Контакты'));
        fireEvent.click(getByText('Интересы'));

        ['work', 'hobby', 'music', 'films', 'books', 'games', 'about'].forEach((id) =>
            matchInput(id)
        );
    });

    it('Меняем образование', () => {
        const {container, getByText} = renderTestContext();
        fireEvent.click(getByText('Образование'));

        const matchInput = (id: string, action?: (input: NodeListOf<Element>) => void) => {
            const inputs = container.querySelectorAll(`[data-id="${id}"]`);
            action && action(inputs);
            expect(inputs).toMatchSnapshot();
        }

        ['name', 'address', 'startDate', 'finishDate', 'specialization'].forEach((id) =>
            matchInput(id, (inputs) =>
                inputs.forEach((input) => fireEvent.change(input, { target: { value: '' }}))
            )
        );

        fireEvent.click(getByText('Контакты'));
        fireEvent.click(getByText('Образование'));

        ['name', 'address', 'startDate', 'finishDate', 'specialization'].forEach((id) =>
            matchInput(id)
        );
    })
});
