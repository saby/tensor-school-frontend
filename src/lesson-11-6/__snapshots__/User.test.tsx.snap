// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Вкладка Интересы 1`] = `
<div>
  <div
    class="User"
  >
    <div
      class="User__fio"
    >
      Иванов Иван Иванович
    </div>
    <div
      class="User__additionalInfo"
    >
      мужской
       
      13.06.1997
    </div>
    <div
      class="User__tabs"
    >
      <div
        class="User__tab "
        data-id="contacts"
      >
        Контакты
      </div>
      <div
        class="User__tab User__tab-selected"
        data-id="interests"
      >
        Интересы
      </div>
      <div
        class="User__tab "
        data-id="education"
      >
        Образование
      </div>
    </div>
    <div
      class="User__info"
    >
      <div
        class="User__row"
      >
        <label>
          Обо мне
        </label>
        <input
          data-id="about"
          type="text"
          value="человек"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Работа
        </label>
        <input
          data-id="work"
          type="text"
          value="Тензор"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Хобби
        </label>
        <input
          data-id="hobby"
          type="text"
          value="реверс-инжиниринг"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Музыка
        </label>
        <input
          data-id="music"
          type="text"
          value="mood music"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Фильмы
        </label>
        <input
          data-id="films"
          type="text"
          value="дюна"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Книги
        </label>
        <input
          data-id="books"
          type="text"
          value="дюна"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Игры
        </label>
        <input
          data-id="games"
          type="text"
          value="Ведьмак"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Context Вкладка Образование 1`] = `
<div>
  <div
    class="User"
  >
    <div
      class="User__fio"
    >
      Иванов Иван Иванович
    </div>
    <div
      class="User__additionalInfo"
    >
      мужской
       
      13.06.1997
    </div>
    <div
      class="User__tabs"
    >
      <div
        class="User__tab "
        data-id="contacts"
      >
        Контакты
      </div>
      <div
        class="User__tab "
        data-id="interests"
      >
        Интересы
      </div>
      <div
        class="User__tab User__tab-selected"
        data-id="education"
      >
        Образование
      </div>
    </div>
    <div
      class="User__info"
    >
      <h3>
        Школа
      </h3>
      <div
        class="User__row"
      >
        <label>
          Название
        </label>
        <input
          data-id="name"
          value="322"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Адрес
        </label>
        <input
          data-id="address"
          value="Ярославль, ул. Пушкина д. Колотушкина"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Дата начала
        </label>
        <input
          data-id="startDate"
          type="date"
          value="2004-09-01"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Дата окончания
        </label>
        <input
          data-id="finishDate"
          type="date"
          value="2015-05-01"
        />
      </div>
      <h3>
        Университет
      </h3>
      <div
        class="User__row"
      >
        <label>
          Название
        </label>
        <input
          data-id="name"
          value="ЯрГУ им.П.Г.Демидова"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Адрес
        </label>
        <input
          data-id="address"
          value="Ярославль, ул. Союзная 144"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Дата начала
        </label>
        <input
          data-id="startDate"
          type="date"
          value="2015-09-01"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Дата окончания
        </label>
        <input
          data-id="finishDate"
          type="date"
          value="2019-08-31"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Специализация
        </label>
        <input
          data-id="specialization"
          value="Компьютерные науки"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Context Вкладка по умолчанию - Контакты 1`] = `
<div>
  <div
    class="User"
  >
    <div
      class="User__fio"
    >
      Иванов Иван Иванович
    </div>
    <div
      class="User__additionalInfo"
    >
      мужской
       
      13.06.1997
    </div>
    <div
      class="User__tabs"
    >
      <div
        class="User__tab User__tab-selected"
        data-id="contacts"
      >
        Контакты
      </div>
      <div
        class="User__tab "
        data-id="interests"
      >
        Интересы
      </div>
      <div
        class="User__tab "
        data-id="education"
      >
        Образование
      </div>
    </div>
    <div
      class="User__info"
    >
      <div
        class="User__row"
      >
        <label>
          Телефон
        </label>
        <input
          data-id="phone"
          type="tel"
          value="+701234567890"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          email
        </label>
        <input
          data-id="email"
          type="email"
          value="myemail@mail.com"
        />
      </div>
      <div
        class="User__row"
      >
        <label>
          Местонахождение
        </label>
        <input
          data-id="location"
          value="г. Ярославль"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Context Меняем интересы 1`] = `
<input
  data-id="work"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 2`] = `
<input
  data-id="hobby"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 3`] = `
<input
  data-id="music"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 4`] = `
<input
  data-id="films"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 5`] = `
<input
  data-id="books"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 6`] = `
<input
  data-id="games"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 7`] = `
<input
  data-id="about"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 8`] = `
<input
  data-id="work"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 9`] = `
<input
  data-id="hobby"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 10`] = `
<input
  data-id="music"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 11`] = `
<input
  data-id="films"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 12`] = `
<input
  data-id="books"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 13`] = `
<input
  data-id="games"
  type="text"
  value=""
/>
`;

exports[`Context Меняем интересы 14`] = `
<input
  data-id="about"
  type="text"
  value=""
/>
`;

exports[`Context Меняем контакты 1`] = `
<input
  data-id="phone"
  type="tel"
  value="+79876543210"
/>
`;

exports[`Context Меняем контакты 2`] = `
<input
  data-id="email"
  type="email"
  value="not_my@email.com"
/>
`;

exports[`Context Меняем контакты 3`] = `
<input
  data-id="location"
  value="где-то"
/>
`;

exports[`Context Меняем контакты 4`] = `
<input
  data-id="phone"
  type="tel"
  value="+79876543210"
/>
`;

exports[`Context Меняем контакты 5`] = `
<input
  data-id="email"
  type="email"
  value="not_my@email.com"
/>
`;

exports[`Context Меняем контакты 6`] = `
<input
  data-id="location"
  value="где-то"
/>
`;

exports[`Context Меняем образование 1`] = `
NodeList [
  <input
    data-id="name"
    value=""
  />,
  <input
    data-id="name"
    value=""
  />,
]
`;

exports[`Context Меняем образование 2`] = `
NodeList [
  <input
    data-id="address"
    value=""
  />,
  <input
    data-id="address"
    value=""
  />,
]
`;

exports[`Context Меняем образование 3`] = `
NodeList [
  <input
    data-id="startDate"
    type="date"
    value=""
  />,
  <input
    data-id="startDate"
    type="date"
    value=""
  />,
]
`;

exports[`Context Меняем образование 4`] = `
NodeList [
  <input
    data-id="finishDate"
    type="date"
    value=""
  />,
  <input
    data-id="finishDate"
    type="date"
    value=""
  />,
]
`;

exports[`Context Меняем образование 5`] = `
NodeList [
  <input
    data-id="specialization"
    value=""
  />,
]
`;

exports[`Context Меняем образование 6`] = `
NodeList [
  <input
    data-id="name"
    value=""
  />,
  <input
    data-id="name"
    value=""
  />,
]
`;

exports[`Context Меняем образование 7`] = `
NodeList [
  <input
    data-id="address"
    value=""
  />,
  <input
    data-id="address"
    value=""
  />,
]
`;

exports[`Context Меняем образование 8`] = `
NodeList [
  <input
    data-id="startDate"
    type="date"
    value=""
  />,
  <input
    data-id="startDate"
    type="date"
    value=""
  />,
]
`;

exports[`Context Меняем образование 9`] = `
NodeList [
  <input
    data-id="finishDate"
    type="date"
    value=""
  />,
  <input
    data-id="finishDate"
    type="date"
    value=""
  />,
]
`;

exports[`Context Меняем образование 10`] = `
NodeList [
  <input
    data-id="specialization"
    value=""
  />,
]
`;
